components/plan-build/bin/hab-plan-build.sh

```diff
@@ -2318,28 +2318,63 @@ SRC_PATH="$PLAN_CONTEXT"
+# confused if they edit one to have this program read the other.
+# The exception to this is when the $pkg_target requires variations to
+# the default `plan.sh`. Plan authors can create these variants by placeing 
+# Plan contents in a subdirectory named the same as the intended $pkg_target. 
+# In this case, we want to check if there is a variant `plan.sh` available 
+# for the $pkg_target, and prefer loading that one. Alternate $pkg_target 
+# are not considered as part of this load logic.
+
+# With plan variants, plans can exist in 4 places per $pkg_target relative to 
+# the `$PLAN_CONTEXT` directory. Only two combinations are allowed:
+#
+#   `./plan.sh` AND `./$pkg_target/plan.sh` 
+#   OR 
+#   `./habitat/plan.sh` AND `./habitat/$pkg_target/plan.sh`
+# Consider all other combination of two plans and abort if found.
+# All combinations of 3 and 4 plans are covered by a 2 plan case, so we don't
+# need to consider those scenarios. 
 
-# We'll make sure that both are not present, and if so abort.
 if [[ -f "$PLAN_CONTEXT/plan.sh" && -f "$PLAN_CONTEXT/habitat/plan.sh" ]];then
   places="$PLAN_CONTEXT/plan.sh and $PLAN_CONTEXT/habitat/plan.sh"
+elif [[ -f "$PLAN_CONTEXT/plan.sh" && -f "$PLAN_CONTEXT/habitat/$pkg_target/plan.sh" ]]; then
+  places="$PLAN_CONTEXT/plan.sh and $PLAN_CONTEXT/habitat/$pkg_target/plan.sh"
+elif [[ -f "$PLAN_CONTEXT/habitat/plan.sh" && -f "$PLAN_CONTEXT/$pkg_target/plan.sh" ]]; then 
+  places="$PLAN_CONTEXT/habitat/plan.sh and $PLAN_CONTEXT/$pkg_target/plan.sh"
+elif [[ -f "$PLAN_CONTEXT/$pkg_target/plan.sh" && -f "$PLAN_CONTEXT/habitat/$pkg_target/plan.sh" ]]; then
+  places="$PLAN_CONTEXT/$pkg_target/plan.sh and $PLAN_CONTEXT/habitat/$pkg_target/plan.sh"
+fi
+
+if [[ -v places ]]; then 
   exit_with "A Plan file was found at $places. Only one is allowed at a time" 42
 fi
-# We check if the provided path has a `plan.sh` in it in either location. If
-# not, we'll quickly bail.
-if [[ ! -f "$PLAN_CONTEXT/plan.sh" ]]; then
-  if [[ -f "$PLAN_CONTEXT/habitat/plan.sh" ]]; then
-    # As the `plan.sh` is in a deeper subdirectory, we'll update the
-    # `$PLAN_CONTEXT` directory to be relative to the actual `plan.sh` file.
-    PLAN_CONTEXT="$PLAN_CONTEXT/habitat"
-  else
-    places="$PLAN_CONTEXT/plan.sh or $PLAN_CONTEXT/habitat/plan.sh"
-    exit_with "Plan file not found at $places" 42
-  fi
+
+build_line "Checking for variant at $pkg_target"
+
+# We check if the provided path has a `plan.sh` in it in valid locations. If
+# not, we'll quickly bail. Prefer most-specific to least specific. 
+if [[ -f "$PLAN_CONTEXT/habitat/$pkg_target/plan.sh" ]]; then
+  PLAN_CONTEXT="$PLAN_CONTEXT/habitat/$pkg_target"
+elif [[ -f "$PLAN_CONTEXT/$pkg_target/plan.sh" ]]; then
+  PLAN_CONTEXT="$PLAN_CONTEXT/$pkg_target"
+elif [[ -f "$PLAN_CONTEXT/habitat/plan.sh" ]]; then 
+  PLAN_CONTEXT="$PLAN_CONTEXT/habitat"
+elif [[ -f "$PLAN_CONTEXT/plan.sh" ]]; then
+	: # DO NOTHING
+else
+  locations=("$PLAN_CONTEXT/habitat/$pkg_target/plan.sh" 
```

Comment:
Might be nice to drive the above if/elif chain from this `locations` array to reduce duplication and make things a little more clear.